### PR TITLE
Engine Command and E2E Testing

### PR DESCRIPTION
**Pull request by ColDog**
This introduces a new command that uses the viper generator. This is the engine command. It is comprised of the 3 components (api, controller and worker). Additionally, end to end testing was introduced to test the engine.

* Introduce commands for each of the components.
* Introduce an 'all' command which runs all of the components together.
* Add an e2e package which runs a large amount of games with contrived latency to test the locking, speed and consistency.
* Fix a race condition with the store, the store will now throw an "invalid sequence" error when a turn is written that is in the wrong sequence.

### CODE REVIEW

**Comment by graeme-hill** [comment]
Cool I guess I wasn't the only one wanting e2e tests.

**Comment by ColDog** [comment]
Ya this is a start I'd say to e2e tests. I didn't really put a ton of effort into devising interesting test cases. This code was more just to test the controller to worker connections a little more comprehensively.

File: api/server.go [hunk 1]
```diff
@@ -160,10 +160,4 @@ func getTicks(w http.ResponseWriter, r *http.Request, ps httprouter.Params, c pb
...
 
 // WaitForExit starts up the server and blocks until the server shuts down.
-func (s *Server) WaitForExit() {
-	log.Infof("Battlesnake engine api listening on %s", s.hs.Addr)
-	err := s.hs.ListenAndServe()
-	if err != nil {
-		log.Errorf("Error while listening: %v", err)
-	}
-}
+func (s *Server) WaitForExit() error { return s.hs.ListenAndServe() }
```
Comment (LouisRitchie) [inline api/server.go]: :raised_hands:

File: e2e/e2e_test.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+						return
+					}
+				}
+
+				spew.Dump(st)
+				t.Errorf("test failed after: %v", time.Duration(waitTicks)*waitInterval)
+			})
+		}
+	}
+}
```
Comment (LouisRitchie) [inline e2e/e2e_test.go]: This test :sweat:

**Pull request merged**